### PR TITLE
[BIOMAGE-2110] - Use selected cell sets in Trajectory Plot

### PR DESCRIPTION
**Pull request by aerlaut**
# Description
Worker filter cell sets to use in trajectory analysis

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-2110

#### Link to staging deployment URL (or set N/A)
N/A

#### Links to any PRs or resources related to this PR
UI - https://github.com/biomage-org/ui/pull/182
API - https://github.com/biomage-org/api/pull/114
Worker - https://github.com/biomage-org/worker/pull/69

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [x] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: python/src/worker/helpers/s3.py [hunk 1]
```diff
@@ -43,6 +43,42 @@ def get_cell_sets(experiment_id):
...
+def get_cell_sets_dict(cell_sets):
+    cell_sets_dict = {}
+
+    for cell_class in cell_sets:
+        if not cell_sets_dict.get(cell_class["key"]):
+          cell_sets_dict[cell_class["key"]] = {
+              "key": cell_class["key"],
+              "name": cell_class["name"],
+              "rootNode": True,
+              "childrenKeys": [],
```
Comment (ivababukova) [inline python/src/worker/helpers/s3.py:55]: this function and the other one -- `get_cell_ids` have nothing to do with S3. Have you considered adding them as part of a different file under `helpers`?

File: r/R/trajectory_analysis.R [hunk 2]
```diff
@@ -5,6 +5,8 @@
     umap = "UMAP"
   )
 
+  TRAJECTORY_ANALYSIS_CACHE <- list()
```
Comment (ivababukova) [inline r/R/trajectory_analysis.R:8]: Can you tell me more details about this `TRAJECTORY_ANALYSIS_CACHE`? How does it work?

File: r/R/trajectory_analysis.R [hunk 3]
```diff
@@ -119,28 +122,27 @@ runTrajectoryAnalysisPseudoTimeTask <- function(req, data) {
...
-  
+
   cell_data <- monocle3::order_cells(cell_data, reduction_method = monocle_embedding_method, root_pr_nodes = root_ids)
 
   pseudotime <- as.data.frame(cell_data@principal_graph_aux@listData$UMAP$pseudotime)
 
   # fill in the NULL values for filtered cells
-  pseudotime <- fillNullForFilteredCells(pseudotime, data)
-
+  pseudotime <- fillNullForFilteredCells(pseudotime, subsetIds(data, req$body$cell_ids))
```
Comment (ivababukova) [inline r/R/trajectory_analysis.R:145]: what is `subsetIds` and why do we need to use it?

File: python/src/tests/data/MockCellSet.json [hunk 4]
```diff
@@ -1,94 +1,97 @@
...
-                "color": "#8c564b",
-                "key": "patient-b",
-                "name": "Patient B"
-            }
-        ]
-    }
-]
\ No newline at end of file
+{
+  "cellSets":
```
Comment (ivababukova) [inline python/src/tests/data/MockCellSet.json:2]: Will the worker receive `cellSets` every time we call trajectory analysis from this PR onwards?

File: r/R/trajectory_analysis.R (hunk 3 above)
Comment (saracastel) [inline r/R/trajectory_analysis.R:145]: Maybe for greater clarity you could first get the subset data and then pass it to the `fillNullForFilteredCells` function.
Something like this: `data_subset <- subsetIds(data, req$body$cell_ids)
pseudotime <- fillNullForFilteredCells(pseudotime, data_subset)`

File: r/R/trajectory_analysis.R [hunk 5]
```diff
@@ -156,9 +158,29 @@ runTrajectoryAnalysisPseudoTimeTask <- function(req, data) {
...
 #' @export
-generateTrajectoryGraph <- function(embedding_data, embedding_settings, clustering_settings, data) {
+generateTrajectoryGraph <- function(
+  embedding_data,
+  embedding_settings,
+  clustering_settings,
+  cell_ids,
+  data
+) {
+
```
Comment (saracastel) [inline r/R/trajectory_analysis.R:168]: Should we pass to the `generateTrajectoryGraph` only req and data, and then extract from req what's needed inside the function? This way when we need to change the arguments as happened in this case, we can modify it only here and not in two places ([here](https://github.com/biomage-org/worker/blob/58691a978a4d79517698994f76cfc835883b9d94/r/R/trajectory_analysis.R#L40-L44), and [here](https://github.com/biomage-org/worker/blob/58691a978a4d79517698994f76cfc835883b9d94/r/R/trajectory_analysis.R#L122-L126))

**Comment by saracastel** [comment]
I tested it locally using this branch + [this UI branch](https://github.com/biomage-org/ui/pull/182) + [this API branch](https://github.com/biomage-org/api/pull/114), but for some reason I'm stuck at the "Calculate root nodes" step. It correctly calculates and shows the root nodes, but then it never calculates the pseudotime. Not sure if it's something wrong on my side. Is it working for you? 
Screenshot 2023-02-22 at 10 24 04

**Review by gerbeldo — APPROVED** [review]

File: python/src/tests/data/MockCellSet.json (hunk 4 above)
Comment (aerlaut) [reply python/src/tests/data/MockCellSet.json:2]: This is the correct shape of the cell sets object (an object with the key `cellSets: [...]`). The previous shape wasn't correct. This change fixes that, so we'll have a mock cell data object which is more similar to the real one.

File: r/R/trajectory_analysis.R (hunk 5 above)
Comment (aerlaut) [reply r/R/trajectory_analysis.R:168]: No, I don't think so. Although it makes the signature of the function cleaner (i.e. less parameters), it'll make the function harder for reuse, and we risk passing information that's not needed by the `generateTrajectoryGraph` via the `req` variable. I think it's ok to pass each of the values, so it's clear what parameters is used by `generateTrajectoryGraph`.

File: r/R/trajectory_analysis.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (aerlaut) [reply r/R/trajectory_analysis.R:8]: I think the solution is pretty simple, and it speeds up calculating the pseudotime. Given the same cell set and trajectory, the user doesn't have to rerun `generateTrajectoryGraph` when calculating pseudotime, saving a lot of time.

[the remark this answers is not in the document]
Comment (aerlaut) [reply r/R/trajectory_analysis.R:8]: The only risk of doing this is if the user changes the cell sets a lot. In that case, the value of `TRAJECTORY_ANALYSIS_CACHE` will be large, and may grow to occupy more than the amount of available memory. I assume this will lead to a crash of the worker pod, which then restarts and clears the working memory.

To handle this, we can be smarter and have something like an LRU cache, or just limit the number of cache keys that can be stored (e.g. 10) evict the oldest cache.

File: r/R/trajectory_analysis.R [hunk 6]
```diff
@@ -193,6 +216,10 @@ generateTrajectoryGraph <- function(embedding_data, embedding_settings, clusteri
 
   cell_data <- monocle3::learn_graph(cell_data)
 
+  TRAJECTORY_ANALYSIS_CACHE[[cache_hash]] <- cell_data
```
[the remark this answers is not in the document]
Comment (aerlaut) [reply r/R/trajectory_analysis.R:219]: Yes

File: r/R/trajectory_analysis.R [hunk 7]
```diff
@@ -178,6 +200,7 @@ generateTrajectoryGraph <- function(embedding_data, embedding_settings, clusteri
   }
 
   data <- assignEmbedding(embedding_data, data)
+  data <- subsetIds(data, cell_ids)
```
[the remark this answers is not in the document]
Comment (aerlaut) [reply r/R/trajectory_analysis.R:203]: Hmm... good idea. We'll need to subset the embedding data too though, before assigning it. Might be tricky?

File: r/R/trajectory_analysis.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (gerbeldo) [reply r/R/trajectory_analysis.R:8]: The cached object size will scale with the number of cells. For big datasets it might get out of hand quickly. What happens after the worker dies? Error screen and a new one has to be spawned?

**Review by saracastel — APPROVED** [review]

File: r/R/trajectory_analysis.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (gerbeldo) [reply r/R/trajectory_analysis.R:8]: I would test with a decent size dataset and see how many trajectories you can calculate before the worker exploding. I'm not sure how users use the feature, and how useful a cache like this might be.

We could just put it in the wild and see how it goes

Comment (ivababukova) [reply r/R/trajectory_analysis.R:8]: Makes sense. If the cell sets change (for example, the user deletes or creates a new Custom cell set), will the cache get invalidated?

File: python/src/tests/data/MockCellSet.json (hunk 4 above)
Comment (ivababukova) [reply python/src/tests/data/MockCellSet.json:2]: why did the unit test pass with an incorrect object?

File: r/R/trajectory_analysis.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivababukova) [reply r/R/trajectory_analysis.R:8]: If we use S3 for cache of worker results, what is the value of trying to implement another type of caching mechanism instead of reusing what we've got?

[the remark this answers is not in the document]
Comment (ivababukova) [reply r/R/trajectory_analysis.R:8]: in my experience, one wants to minimise the creation of "custom snowflakes", because these are just creators for corner cases. And corner cases eat up so much maintenance and development time

Comment (aerlaut) [reply r/R/trajectory_analysis.R:8]: yes, if the use uses the newly created cell set for trajectory analysis.

File: python/src/tests/data/MockCellSet.json (hunk 4 above)
Comment (aerlaut) [reply python/src/tests/data/MockCellSet.json:2]: No, I fixed the test to use the new format. You can see the changes in these files:
- [python/src/tests/helpers/test_find_cell_ids_in_same_hierarchy.py](https://github.com/biomage-org/worker/pull/69/files#diff-65423df5c1fedbafc51c516d16706b52e52a95adc38896965acecb264b3f2cd2)
- [python/src/tests/helpers/test_get_diff_expr_cellsets.py](https://github.com/biomage-org/worker/pull/69/files#diff-9d3f0da8d51cacc72afb5c30cad33444a05f7a3bb2e05411abe1020da8402c67)

File: r/R/trajectory_analysis.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (aerlaut) [reply r/R/trajectory_analysis.R:8]: @ivababukova this is not a result that we want to display to the user. It'll never be fetched by the UI, so I don't think we need to store it in S3. It's an intermediate value that is used during the calculation of trajectory analysis.

[the remark this answers is not in the document]
Comment (aerlaut) [reply r/R/trajectory_analysis.R:8]: @gerbeldo to my knowledge, the pod will just restart. There will be an error message in Lens saying that the pod crashes, but that should be it. To make this safer and not go out of proportion, I can make the cache simpler (e.g. just store the value of 1 calculation), so it'll never go too big. That will be good enough as it'll shorten down the calculation time of the pseudotime.

[the remark this answers is not in the document]
Comment (gerbeldo) [reply r/R/trajectory_analysis.R]: And the user sees a "try again" button or something like that. Ok, makes sense. I'm not convinced of the usecase, and I fear that this will more often than not just kill the worker for bigger experiments after a few calculations.

File: r/R/trajectory_analysis.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (aerlaut) [reply r/R/trajectory_analysis.R]: turns out it can be done out of the box

File: r/R/trajectory_analysis.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (aerlaut) [reply r/R/trajectory_analysis.R]: I have changed this to only store the result of only one experiment,so it'll not use more memory if more trajectory analysis are done.
